### PR TITLE
extractors: Add dtype validation for autogenerated extractors

### PR DESCRIPTION
**Pull request by aditya-nambiar**

### CODE REVIEW

**Review by nikhilgarg28 — APPROVED** [review]

File: fennel/featuresets/test_derived_extractors.py [hunk 1]
```diff
@@ -57,7 +57,7 @@ class UserInfo:
             default="unspecified",
         )
         # lookup with meta
-        age_years: float = (
+        age_years: int = (
```
Comment (zsid60) [inline fennel/featuresets/test_derived_extractors.py:60]: In the dataset age is defined as a float so why this change?

File: fennel/featuresets/test_invalid_featureset.py [hunk 2]
```diff
@@ -238,3 +245,114 @@ class UserInfo:
...
+    # age should be optional[int]
+    if sys.version_info < (3, 9):
+        assert (
+            str(e.value)
+            == "Feature `UserInfo.age` has type `<class 'int'>` but expectected type `typing.Union[int, NoneType]`"
+        )
+    else:
+        assert (
+            str(e.value)
+            == "Feature `UserInfo.age` has type `<class 'int'>` but expectected type `typing.Optional[int]`"
```
Comment (zsid60) [inline fennel/featuresets/test_invalid_featureset.py:283]: I'm a bit confused on the point of this PR. In this example the dataset field is non optional and so intuitively to me the extracted feature would also be non optional. For values where the lookup doesn't return anything, this is what the default is for. So if you'll always have an int from the ds or a default int, what is this automatic relaxing to an optional type giving us / why are we doing it? Is it for the case when there is no default value (if so, why is this generalized to even when a default is present?)

File: fennel/featuresets/featureset.py [hunk 3]
```diff
@@ -576,6 +582,60 @@ def _get_extractors(self) -> List[Extractor]:
         return extractors
 
     def _validate(self):
+        cls_module = inspect.getmodule(self.__fennel_original_cls__)
+        if cls_module is not None and hasattr(
+            cls_module, FENNEL_GEN_CODE_MARKER
```
Comment (zsid60) [inline fennel/featuresets/featureset.py:587]: What does this do?

File: fennel/featuresets/test_derived_extractors.py (hunk 1 above)
Comment (aditya-nambiar) [reply fennel/featuresets/test_derived_extractors.py:60]: No its a float

File: fennel/featuresets/test_invalid_featureset.py (hunk 2 above)
Comment (aditya-nambiar) [reply fennel/featuresets/test_invalid_featureset.py:283]: I dont get your question, this example does not have a default value, and hence age should be Optional[int], if you specify a default it should be in

File: fennel/featuresets/featureset.py (hunk 3 above)
Comment (aditya-nambiar) [reply fennel/featuresets/featureset.py:587]: This is an interesting issue I encountered. So basically we ship code. But when the shipped code runs, since it runs under exec mode, all types are strings. Hence for the shipping code typing validation breaks. However we dont need to run validation once it has already run by the user, so we disable the validation when executing shipped code

Comment (zsid60) [reply fennel/featuresets/featureset.py:587]: Ahh I see

File: fennel/featuresets/test_derived_extractors.py (hunk 1 above)
Comment (zsid60) [reply fennel/featuresets/test_derived_extractors.py:60]: Oh sorry I was looking at the User dataset not UserInfo
